### PR TITLE
Feat/gdn decode pooled

### PR DESCRIPTION
**Pull request by xutizhou**
## 📌 Description

This PR adds pool-indexed (indirect) state access to the GDN decode kernel, enabling zero-copy integration with SGLang's state pool architecture.

### Background: SGLang's State Pool Architecture

In SGLang, when serving linear attention models (like Qwen3-Next using Gated Delta Rule), we maintain a **state pool** to store recurrent states for all active requests:

`ssm_states: [num_layers, pool_size, num_heads, head_dim, head_dim]`

where `pool_size` = `max_num_reqs` (maximum concurrent requests).

Each active request has a `req_pool_idx` that maps it to a slot in this pool. The mapping is **not contiguous** - requests come and go, so indices can be scattered (e.g., a batch of 4 requests might have pool indices `[3, 7, 12, 25]`).

### Motivation

The current GDN decode kernel expects state with shape `[B, H, K, V]` where B equals batch size and there's a 1:1 mapping (batch index i → state index i). To use it with SGLang's pool, we would need to:

1. **Gather** states from pool indices before kernel call
2. Run kernel on contiguous `[B, H, K, V]` state  
3. **Scatter** updated states back to pool indices

This adds 2 extra memory copy operations per decode step.

### Changes

This PR adds a `state_indices` parameter for **zero-copy pool access**:

```python
def gated_delta_rule_decode_pretranspose(
    q, k, v, beta,
    state,           # Can be [pool_size, H, K, V] instead of [B, H, K, V]
    state_indices,   # NEW: int32 tensor [B] mapping batch_idx -> pool_idx
    ...
)
```

When `state_indices` is provided:
- Kernel uses indirect addressing: `state[state_indices[batch_idx]]` instead of `state[batch_idx]`
- Negative indices (padding slots for CUDA graph) skip computation and write zeros to output
- Eliminates gather/scatter overhead + host-side `torch.where` for padding (~37μs/call)

## 🔍 Related Issues

- [sgl-project/sglang#18361](https://github.com/sgl-project/sglang/pull/18361) - FlashInfer K-last GDN integration into SGLang

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Reviewer Notes

This PR is required for integrating FlashInfer's K-last GDN kernels into SGLang. The pool indexing feature allows SGLang to directly use its state pool without gather/scatter overhead.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Optional pooled (indirect) state access via a new state_indices parameter enabling zero-copy pooled state handling; negative indices yield zeroed outputs.

* **Improvements**
  * Kernel launch paths, grid sizing, and compiled-kernel caching differentiate pooled vs. non-pooled modes; APIs and docstrings updated to propagate pooling flags while maintaining compatibility.

* **Tests**
  * New test suite validating pooled decode correctness, padding/negative-index behavior, state updates, and pooled vs. non-pooled equivalence.

### CODE REVIEW

**Comment by yzh119** [comment]
@xutizhou can you explain what this PR is about? e.g. adding descriptions about what pool is in GDN.

**Comment by xutizhou** [comment]
> @xutizhou can you explain what this PR is about? e.g. adding descriptions about what pool is in GDN.

updated in the description.

**Comment by yzh119** [comment]
Hi @xutizhou can you also port your work to the f16 state kernel that have already been merged in #2498 ?

**Comment by xutizhou** [comment]
> Hi @xutizhou can you also port your work to the f16 state kernel that have already been merged in #2498 ?

in this pr or open a new pr?

**Comment by xutizhou** [comment]
> Hi @xutizhou can you also port your work to the f16 state kernel that have already been merged in #2498 ?

Ported in a new pr #2634

File: flashinfer/gdn_decode.py [hunk 1]
```diff
@@ -231,189 +232,208 @@ def gdn_decode_kernel_small_batch_pretranspose(
...
+    # When pool indexing: skip computation for padding slots (negative indices)
+    # and write zeros to output. When not pool indexing: always valid (pool_idx=0).
+    if pool_idx >= 0:
+        # Get current batch
+        gSrc_batch = h0_source[(state_idx, None, None)]  # (V, K)
+        gDst = cute.local_tile(h0_source, (1, TILE_V, TILE_K), (state_idx, None, 0))
 
-    # Partition for load
-    thr_copy_load = tiled_copy_load.get_slice(tidx)
+        # V 方向分 tiles
```
Comment (hlu1) [inline flashinfer/gdn_decode.py]: Please fix

File: flashinfer/gdn_decode.py [hunk 2]
```diff
@@ -1073,18 +1131,35 @@ def gated_delta_rule_decode_pretranspose(
...
     # Compile kernel with TVM FFI (cached)
-    cache_key = (B, T, H, HV, K, V, q.dtype, scale, use_qk_l2norm)
+    cache_key = (
+        B,
+        T,
+        H,
+        HV,
+        K,
+        V,
+        pool_size,
```
Comment (yzh119) [inline flashinfer/gdn_decode.py]: why do we need pool_size as part of the cache_key? I suppose pool_size could be a runtime parameter instead of compile time?

Comment (xutizhou) [reply flashinfer/gdn_decode.py]: removed pool_size from cache_key.

File: flashinfer/gdn_decode.py (hunk 1 above)
Comment (xutizhou) [reply flashinfer/gdn_decode.py]: done.

File: flashinfer/gdn_decode.py [hunk 3]
```diff
@@ -273,8 +278,18 @@ def gated_delta_rule_decode_pretranspose(
         # Kernel outputs bfloat16, allocate in that dtype first
         output = torch.zeros((B, T, HV, V), dtype=torch.bfloat16, device=q.device)
 
-    # Convert state from [B, HV, V, K] to [B*HV, V, K] for kernel
-    h0_source = state.reshape(B * HV, V, K)
+    # Build h0_source: [pool_size*HV, V, K] for kernel
+    if use_pool:
+        pool_size = initial_state.shape[0]
+        assert initial_state.is_contiguous(), (
```
Comment (ZJY0516) [inline flashinfer/gdn_decode.py:284]: Do we consider to support non-contiguous state?

Comment (xutizhou) [reply flashinfer/gdn_decode.py:284]: in which situation do we need non-contiguous state?

Comment (ZJY0516) [reply flashinfer/gdn_decode.py:284]: vLLM uses non-contiguous state

Comment (xutizhou) [reply flashinfer/gdn_decode.py:284]: > vLLM uses non-contiguous state

For non-contiguous states, we should be able to compute the true indices using strides. Once the assert is removed, it can also work with our kernel.
